### PR TITLE
Synapse Integration

### PR DESCRIPTION
**Pull request by fantix**
### New Features
- Support Synapse as new AuthN and AuthZ provider
- Support AuthZ Provider feature in Arborist

### Dependency updates
- Requires Arborist x.x.x

### Deployment changes
- Requires manual work in Arborist DB to fill in the empty `authz_provider` fields

### CODE REVIEW

File: fence/blueprints/login/__init__.py [hunk 1]
```diff
@@ -75,11 +77,19 @@ def absolute_login_url(provider_id):
...
+                    "url": None,
+                    "desc": None,
+                    "secondary": False,
+                }
             return {
                 "id": idp_id,
                 "name": idps[idp_id]["name"],
                 "url": absolute_login_url(idp_id),
+                "desc": idps[idp_id].get("desc", None),
+                "secondary": idps[idp_id].get("secondary", False),
```
Comment (vpsx) [inline fence/blueprints/login/__init__.py]: desc is text below button
secondary = False --> primary color button

File: fence/models.py [hunk 2]
```diff
@@ -698,6 +698,34 @@ def migrate(driver):
...
+        ELSIF (TG_OP = 'UPDATE') THEN
+            INSERT INTO user_audit_logs (timestamp, operation, old_values, new_values)
+            SELECT now(), 'UPDATE', row_to_json(OLD), row_to_json(NEW);
+            RETURN NEW;
+        ELSIF (TG_OP = 'INSERT') THEN
+            INSERT INTO user_audit_logs (timestamp, operation, new_values)
+            SELECT now(), 'INSERT', row_to_json(NEW);
+            RETURN NEW;
+        END IF;
+        RETURN NULL;
```
[the remark this answers is not in the document]
Comment (fantix) [reply fence/models.py]: Yeah it works without `RETURN NULL;` - however this was copied from PostgreSQL docs, I suppose the reason here is to honor the function definition.

File: fence/models.py [hunk 3]
```diff
@@ -698,6 +698,34 @@ def migrate(driver):
         metadata=md,
     )
 
+    with driver.session as session:
+        session.execute("""\
+CREATE OR REPLACE FUNCTION process_user_audit() RETURNS TRIGGER AS $user_audit$
+    BEGIN
+        IF (TG_OP = 'DELETE') THEN
+            INSERT INTO user_audit_logs (timestamp, operation, old_values)
```
[the remark this answers is not in the document]
Comment (fantix) [reply fence/models.py]: Yeah we have at least the migration in this repo - sry what was your stand here again pls?

File: fence/sync/sync_users.py [hunk 4]
```diff
@@ -1073,15 +1073,17 @@ def _update_arborist(self, session, user_yaml):
 
         policies = user_yaml.rbac.get("policies", [])
         for policy in policies:
+            policy_id = policy.pop("id")
+            if not policy_id:
+                continue
```
[the remark this answers is not in the document]
Comment (fantix) [reply fence/sync/sync_users.py]: Right it shouldn't happen - fixed!

File: fence/sync/sync_users.py [hunk 5]
```diff
@@ -1073,15 +1073,17 @@ def _update_arborist(self, session, user_yaml):
...
             try:
-                response = self.arborist_client.create_policy(
-                    policy, skip_if_exists=True
-                )
-                if response:
-                    self._created_policies.add(policy["id"])
+                response = self.arborist_client.update_policy(policy_id, policy)
```
[the remark this answers is not in the document]
Comment (fantix) [reply fence/sync/sync_users.py]: Yes! Thanks!

File: fence/blueprints/login/__init__.py (hunk 1 above)
Comment (fantix) [reply fence/blueprints/login/__init__.py]: *above* button :P but ya

File: fence/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vpsx) [reply fence/models.py]: No sorry this was just a note haha

**Pull request merged**